envisage/ids.py

```diff
@@ -0,0 +1,62 @@
+# (C) Copyright 2007-2019 Enthought, Inc., Austin, TX
+# All rights reserved.
+#
+# This software is provided without warranty under the terms of the BSD
+# license included in LICENSE.txt and may be redistributed only
+# under the conditions described in the aforementioned license.  The license
+# is also available online at http://www.enthought.com/licenses/BSD.txt
+# Thanks for using Enthought open source!
+
+from envisage.core_plugin import CorePlugin
```

Comment:
What's the impact of these imports?

I think we want any importing from `envisage.ids` to be as lightweight as possible. In particular, importing `envisage.ids` should:

- not cause any import-time side-effects
- not require any optional dependencies
- be fast

Duplicating the id strings rather than getting them directly from the plugins may be the lesser of the two evils here.

Reply:
For more esoteric things (like the `ProjectPlugin`) there's also the option of having an `ids` module at some other level in the codebase. But I'd suggest not doing that unless a real need emerges.

Reply:
Importing from `envisage.ids` (after removing `ProjectPlugin` and `WorkbenchPlugin`) has

- no import-time side-effects
- does not require any optional dependencies (all of the optional dependencies e.g. `ipykernel`) are hidden inside plugin methods.
- importing from `envisage.ids` should definitely be faster than importing from `envisage.api` but it will probably be faster if we redefine the ID strings.

Reply:
Okay, thanks. I do see a noticeable lag on importing `envisage.ids`, though, and the list of modules imported as a result is large. I think we'd do better to go with duplicated ID strings.

```
Python 3.8.1 (default, Dec 22 2019, 03:45:23) 
[Clang 10.0.1 (clang-1001.0.46.4)] on darwin
Type "help", "copyright", "credits" or "license" for more information.
>>> import sys
>>> mods1 = sys.modules.copy()
>>> import envisage
>>> mods2 = sys.modules.copy()
>>> import envisage.ids
>>> mods3 = sys.modules.copy()
>>> set(mods3) - set(mods2)
{'_random', 'dis', 'email.errors', '_osx_support', '_ssl', 'traits.adaptation.adaptation_manager', 'xml.parsers.expat', 'pkg_resources.extern.six', 'envisage.application_event', 'signal', 'plistlib', 'pkg_resources._vendor', 'traits._py2to3', 'envisage.unknown_extension', '_uuid', '_compression', 'envisage.import_manager', 'zipfile', 'traits.etsconfig.api', 'select', 'pkg_resources._vendor.packaging.__about__', 'fnmatch', 'envisage.egg_plugin_manager', 'envisage.plugins', 'apptools._version', 'envisage.ui', '_pickle', 'envisage.extension_point_binding', 'importlib._bootstrap_external', 'traits.has_traits', 'pkg_resources.extern.six.moves.urllib', 'traits', 'urllib.request', '_sha512', 'pickle', 'traits.adaptation.adaptation_error', 'email._parseaddr', 'email.charset', 'grp', 'envisage.plugin_extension_registry', 'ntpath', '_opcode', 'urllib.error', 'xml', 'traits.adaptation.adapter', 'email.feedparser', 'envisage.i_provider_extension_registry', 'traits.trait_base', '_socket', 'http', '_sha3', 'envisage.plugin_activator', 'socket', 'traits.util.toposort', 'pkg_resources.py31compat', 'envisage.unknown_extension_point', 'email.message', 'traits.etsconfig', '_scproxy', 'envisage.ui.tasks', 'pkg_resources._vendor.six.moves', 'envisage.i_plugin_activator', '_hashlib', 'urllib.response', 'platform', 'envisage.i_application', 'pyface._version', 'envisage.extension_registry', 'traitsui._version', 'envisage.ui.tasks.preferences_category', 'locale', 'email.encoders', 'contextlib', 'envisage.plugin_event', 'envisage.application', 'envisage.extension_provider', 'envisage.plugin_manager', 'traits.api', 'traits.traits', 'envisage.i_extension_point_user', 'pkg_resources', 'pyexpat.model', 'envisage.plugins.python_shell.python_shell_plugin', 'apptools.preferences.package_globals', 'envisage.i_plugin', 'envisage._compat', 'envisage.ui.tasks.tasks_plugin', 'lzma', 'urllib', 'selectors', 'pkg_resources.extern.appdirs', 'apptools', 'hashlib', 'traits.adaptation.api', 'email.utils', '_lzma', 'quopri', 'sysconfig', 'pkg_resources.extern', 'envisage.i_plugin_manager', 'traits.adaptation', 'apptools.preferences.api', 'envisage.service_offer', '_struct', 'envisage.i_extension_registry', 'pkg_resources.extern.packaging', 'envisage.provider_extension_registry', 'traits.trait_notifiers', 'email.quoprimime', 'email.parser', 'pkg_resources.extern.six.moves', '_sysconfigdata__darwin_darwin', 'envisage.plugins.ipython_kernel.ipython_kernel_plugin', 'envisage.api', 'six.moves', 'traits.trait_types', 'traits.util.api', '_datetime', 'importlib._bootstrap', 'tempfile', 'pwd', 'apptools.preferences.preferences', '_posixsubprocess', 'envisage.egg_utils', 'pyexpat.errors', 'pkg_resources._vendor.six', 'shutil', 'envisage.class_load_hook', 'uuid', 'traits.util', 'zlib', 'email', 'email._policybase', 'xml.parsers.expat.errors', 'traits.util.camel_case', 'pkg_resources.extern.packaging.requirements', 'envisage.i_service_registry', 'envisage.i_import_manager', 'traits.util.deprecated', '_blake2', 'traits.version', 'pkg_resources.extern.packaging.markers', 'http.client', 'envisage.extension_point_changed_event', 'pyexpat', 'traitsui.util', 'subprocess', 'importlib.abc', 'pkgutil', 'traits.category', 'traits.trait_errors', 'envisage.extension_point', 'pkg_resources.extern.packaging._structures', 'base64', 'struct', 'textwrap', 'importlib.machinery', 'traits.util.event_tracer', 'apptools.preferences.scoped_preferences', 'bz2', 'pkg_resources.extern.pyparsing', 'pkg_resources.extern.packaging._compat', 'envisage.core_plugin', 'inspect', 'traitsui', 'envisage.ids', '_bisect', 'apptools.preferences.preference_binding', '_compat_pickle', 'random', 'envisage.i_service_user', 'envisage.plugin', 'apptools.preferences.preferences_helper', '_bz2', 'email._encoded_words', 'envisage.twisted_application', 'math', 'calendar', 'importlib.util', 'envisage.i_extension_provider', 'envisage.i_extension_point', 'bisect', 'pkg_resources.extern.packaging.specifiers', 'envisage.plugins.ipython_kernel', 'email.iterators', 'pyface', 'traitsui.view_element', 'pyface.ui_traits', 'email.header', 'xml.parsers.expat.model', 'importlib', 'urllib.parse', 'traitsui.ui_traits', 'apptools.preferences.i_preferences', 'envisage.service', 'traitsui.view_elements', 'traits.trait_value', 'binascii', 'apptools.preferences', 'errno', 'six', 'traits.adaptation.adaptation_offer', 'uu', 'envisage.plugins.python_shell', 'envisage.safeweakref', 'traits.trait_handlers', 'copy', 'traits.trait_numeric', 'opcode', 'ssl', 'traits.ctraits', 'traits.etsconfig.etsconfig', 'pprint', 'envisage.service_registry', 'email.base64mime', 'pkg_resources.extern.packaging.version', 'datetime', 'xml.parsers', 'traits.util.import_symbol'}
```

Reply:
Sorry, here's a more readable version of that diff:
```
>>> pprint(set(mods3) - set(mods2))
{'_bisect',
 '_blake2',
 '_bz2',
 '_compat_pickle',
 '_compression',
 '_datetime',
 '_hashlib',
 '_lzma',
 '_opcode',
 '_osx_support',
 '_pickle',
 '_posixsubprocess',
 '_random',
 '_scproxy',
 '_sha3',
 '_sha512',
 '_socket',
 '_ssl',
 '_struct',
 '_sysconfigdata__darwin_darwin',
 '_uuid',
 'apptools',
 'apptools._version',
 'apptools.preferences',
 'apptools.preferences.api',
 'apptools.preferences.i_preferences',
 'apptools.preferences.package_globals',
 'apptools.preferences.preference_binding',
 'apptools.preferences.preferences',
 'apptools.preferences.preferences_helper',
 'apptools.preferences.scoped_preferences',
 'base64',
 'binascii',
 'bisect',
 'bz2',
 'calendar',
 'contextlib',
 'copy',
 'datetime',
 'dis',
 'email',
 'email._encoded_words',
 'email._parseaddr',
 'email._policybase',
 'email.base64mime',
 'email.charset',
 'email.encoders',
 'email.errors',
 'email.feedparser',
 'email.header',
 'email.iterators',
 'email.message',
 'email.parser',
 'email.quoprimime',
 'email.utils',
 'envisage._compat',
 'envisage.api',
 'envisage.application',
 'envisage.application_event',
 'envisage.class_load_hook',
 'envisage.core_plugin',
 'envisage.egg_plugin_manager',
 'envisage.egg_utils',
 'envisage.extension_point',
 'envisage.extension_point_binding',
 'envisage.extension_point_changed_event',
 'envisage.extension_provider',
 'envisage.extension_registry',
 'envisage.i_application',
 'envisage.i_extension_point',
 'envisage.i_extension_point_user',
 'envisage.i_extension_provider',
 'envisage.i_extension_registry',
 'envisage.i_import_manager',
 'envisage.i_plugin',
 'envisage.i_plugin_activator',
 'envisage.i_plugin_manager',
 'envisage.i_provider_extension_registry',
 'envisage.i_service_registry',
 'envisage.i_service_user',
 'envisage.ids',
 'envisage.import_manager',
 'envisage.plugin',
 'envisage.plugin_activator',
 'envisage.plugin_event',
 'envisage.plugin_extension_registry',
 'envisage.plugin_manager',
 'envisage.plugins',
 'envisage.plugins.ipython_kernel',
 'envisage.plugins.ipython_kernel.ipython_kernel_plugin',
 'envisage.plugins.python_shell',
 'envisage.plugins.python_shell.python_shell_plugin',
 'envisage.provider_extension_registry',
 'envisage.safeweakref',
 'envisage.service',
 'envisage.service_offer',
 'envisage.service_registry',
 'envisage.twisted_application',
 'envisage.ui',
 'envisage.ui.tasks',
 'envisage.ui.tasks.preferences_category',
 'envisage.ui.tasks.tasks_plugin',
 'envisage.unknown_extension',
 'envisage.unknown_extension_point',
 'errno',
 'fnmatch',
 'grp',
 'hashlib',
 'http',
 'http.client',
 'importlib',
 'importlib._bootstrap',
 'importlib._bootstrap_external',
 'importlib.abc',
 'importlib.machinery',
 'importlib.util',
 'inspect',
 'locale',
 'lzma',
 'math',
 'ntpath',
 'opcode',
 'pickle',
 'pkg_resources',
 'pkg_resources._vendor',
 'pkg_resources._vendor.packaging.__about__',
 'pkg_resources._vendor.six',
 'pkg_resources._vendor.six.moves',
 'pkg_resources.extern',
 'pkg_resources.extern.appdirs',
 'pkg_resources.extern.packaging',
 'pkg_resources.extern.packaging._compat',
 'pkg_resources.extern.packaging._structures',
 'pkg_resources.extern.packaging.markers',
 'pkg_resources.extern.packaging.requirements',
 'pkg_resources.extern.packaging.specifiers',
 'pkg_resources.extern.packaging.version',
 'pkg_resources.extern.pyparsing',
 'pkg_resources.extern.six',
 'pkg_resources.extern.six.moves',
 'pkg_resources.extern.six.moves.urllib',
 'pkg_resources.py31compat',
 'pkgutil',
 'platform',
 'plistlib',
 'pprint',
 'pwd',
 'pyexpat',
 'pyexpat.errors',
 'pyexpat.model',
 'pyface',
 'pyface._version',
 'pyface.ui_traits',
 'quopri',
 'random',
 'select',
 'selectors',
 'shutil',
 'signal',
 'six',
 'six.moves',
 'socket',
 'ssl',
 'struct',
 'subprocess',
 'sysconfig',
 'tempfile',
 'textwrap',
 'traits',
 'traits._py2to3',
 'traits.adaptation',
 'traits.adaptation.adaptation_error',
 'traits.adaptation.adaptation_manager',
 'traits.adaptation.adaptation_offer',
 'traits.adaptation.adapter',
 'traits.adaptation.api',
 'traits.api',
 'traits.category',
 'traits.ctraits',
 'traits.etsconfig',
 'traits.etsconfig.api',
 'traits.etsconfig.etsconfig',
 'traits.has_traits',
 'traits.trait_base',
 'traits.trait_errors',
 'traits.trait_handlers',
 'traits.trait_notifiers',
 'traits.trait_numeric',
 'traits.trait_types',
 'traits.trait_value',
 'traits.traits',
 'traits.util',
 'traits.util.api',
 'traits.util.camel_case',
 'traits.util.deprecated',
 'traits.util.event_tracer',
 'traits.util.import_symbol',
 'traits.util.toposort',
 'traits.version',
 'traitsui',
 'traitsui._version',
 'traitsui.ui_traits',
 'traitsui.util',
 'traitsui.view_element',
 'traitsui.view_elements',
 'urllib',
 'urllib.error',
 'urllib.parse',
 'urllib.request',
 'urllib.response',
 'uu',
 'uuid',
 'xml',
 'xml.parsers',
 'xml.parsers.expat',
 'xml.parsers.expat.errors',
 'xml.parsers.expat.model',
 'zipfile',
 'zlib'}
```